Changes to build and run Titan modules and applications with stock Lua (without changes needed to `luaconf.h`)

mascarenhas opened the pull request:
Link Titan modules statically with part of the Lua object files (only change over a regular Lua installation is that they need to be compiled with `-fPIC`, but no changes to `luaconf.h` are needed.

All `static` functions that every Titan module/app uses should be in a separate `titan.c` file (some of these functions may be compiled from Lua and specialized, with proper attribution).

Titan applications do not need Lua object files compiled with `-fPIC`, for now we can just have one set of object files, in the future we might want to have two.

I am also trying to get Travis to do OSX builds of Titan, but as I do not have a local OSX machine myself I will leave this out of this PR if it turns out to be too much trouble.

mascarenhas committed:
build and load Titan modules with stock Lua

mascarenhas committed:
hiccup, while trying to enable Travis OSX builds ended up trying to run LuaJIT on OSX

mascarenhas committed:
fix language for OSX in Travis

mascarenhas committed:
move common static functions for all Titan modules to Titan runtime

hishamhm commented:
> I am also trying to get Travis to do OSX builds of Titan, but as I do not have a local OSX machine myself I will leave this out of this PR if it turns out to be too much trouble.

I'm going to push a commit here attempting to fix this, I checked the changes against the LuaRocks `.travis.yml`.

hishamhm committed:
updates to .travis.yml

* Travis calls macOS "osx", Lua calls it "macosx"
* Cache the lua_install directory for hererocks-installed build dependencies
* Re-enable coverage reporting and print a summary in the Travis log

mascarenhas commented:
Something pretty odd just happened with this branch, it is showing the latest commits that were already merged into `master`.

hishamhm commented:
@mascarenhas my bad, just refresh :) I messed it up for a second when force-pushing a fixed-fix.

mascarenhas commented:
We can already remove the `lua` folder from the repository, by the way. About this PR, I still have not tested if Titan executables are building correctly (I suspect not), and after fixing this we can merge and do the specialization of `getgeneric` some other time.

mascarenhas commented:
Looks like the FFI needs some fixes to work on OSX. We should just leave the Travis build commented out for now and try to do these fixes in  another PR, preferably with access to an OSX machine.

hishamhm commented:
All right, macOS ran, but c-parser is failing to parse some standard library headers there (those are usually the ones that push the parser the hardest).

hishamhm commented:
@mascarenhas agreed!

hishamhm committed:
disable macOS on Travis for now

The c-parser still needs work with the macOS standard library
headers (these are usually the headers that push the parser
the hardest).

Will reenable once those quirks are addressed.

hishamhm commented on titan-runtime/titan.h:
```diff
@@ -0,0 +1,58 @@
+LUAI_FUNC const TValue* getgeneric(Table* t, const TValue* key);
+LUAI_FUNC void *loadlib (lua_State *L, const char *file);
+LUAI_FUNC void *loadsym (lua_State *L, void *lib, const char *sym);
+LUAI_FUNC const TValue *getgeneric (Table *t, const TValue *key);
```
getgeneric is duplicated here

mascarenhas replied:
oops, pushing a fix

mascarenhas committed:
remove duplicate getgeneric sig

mascarenhas committed:
remove embedded Lua sources, we now build and run against stock Lua

mascarenhas commented:
Ok, will make sure we can still compile Titan applications, and not just Lua modules, later.

mascarenhas committed:
remove unneeded Lua headers from generated code

mascarenhas committed:
make sure we can still build and run applications, and add a test driver for apps

mascarenhas committed:
add test driver for apps with modules

mascarenhas committed:
add include dirs to app compile command so it will build on Travis

hishamhm reviewed (approved):
@mascarenhas last missing bit: could you update the build/install instructions in README.md? I'll leave the PR approved so you can merge once that is in!

mascarenhas committed:
update build and install instructions
